File: language/compiler/src/com/kubadziworski/parsing/visitor/statement/PrintStatementVisitor.java
package com.kubadziworski.parsing.visitor.statement;

import jiva.parser.JivaBaseVisitor;
import jiva.parser.JivaParser;
import jiva.parser.JivaParser.ExpressionContext;
import jiva.parser.JivaParser.PrintStatementContext;
import com.kubadziworski.domain.node.expression.Expression;
import com.kubadziworski.domain.node.statement.PrintStatement;
import com.kubadziworski.parsing.visitor.expression.ExpressionVisitor;
import org.antlr.v4.runtime.misc.NotNull;

public class PrintStatementVisitor extends JivaBaseVisitor<PrintStatement> {
    private final ExpressionVisitor expressionVisitor;

    public PrintStatementVisitor(ExpressionVisitor expressionVisitor) {
        this.expressionVisitor = expressionVisitor;
    }

    @Override
    public PrintStatement visitPrintStatement(@NotNull PrintStatementContext ctx) {
        ExpressionContext expressionCtx = ctx.expression();
        Expression expression = expressionCtx.accept(expressionVisitor);
        return new PrintStatement(expression);
    }
}

File: language/compiler/src/parsing/visitor/statement/PrintStatementVisitor.kt
package jiva.parsing.visitor.statement

import jiva.domain.node.statement.PrintStatement
import jiva.parsing.visitor.expression.ExpressionVisitor
import jiva.parser.JivaBaseVisitor
import jiva.parser.JivaParser.PrintStatementContext

class PrintStatementVisitor(private val visitor: ExpressionVisitor) : JivaBaseVisitor<PrintStatement>() {

    override fun visitPrintStatement(ctx: PrintStatementContext): PrintStatement {
        val expressionCtx = ctx.expression()
        val expression = expressionCtx.accept(visitor)!!
        return PrintStatement(expression)
    }

}